### ISSUE TITLE
Support `.requires_grad_(True)` in `torch.compile`

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#170487]
### 🚀 The feature, motivation and pitch

Currently it fails, no matter how I try to hack around:
- https://github.com/pytorch/pytorch/issues/167729#issuecomment-3657986764

This is often needed to do `torch.autograd.grad` calls inside `torch.compile`, in order to satisfy its requirements (it doesn't like to compute gradient wrt input when input doesn't have requires_grad, although in theory maybe it could also handle this case, because input-for-which-to-compute-grad is explicitly specified anyway)

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo

**Comment by Skylion007** [comment]
Why not also access:
   def dims(self) -> Sequence[DVar | int | _DynType]:
        return self.__args__

**Comment by jansel** [comment]
My agent was unable to reproduce this on 04d232f35374e0e7216d8c4a5d8ba2f6d86bb332. Reopen if you see this on the latest PyTorch nightly

**Comment by vadimkantorov** [comment]
@jansel Do you mean that this snippet now passes?

```python
import torch

@torch.compile(fullgraph=True)
def model_grad(model, x):
    with torch.enable_grad():
        x = x.detach().requires_grad_(True)
        grad = torch.autograd.grad(model(x).sum(), x, create_graph=False)[0]
    return grad.detach()

if __name__ == '__main__':
    model = torch.nn.Sequential(torch.nn.Linear(10, 20), torch.nn.Linear(20, 30))
    grad = model_grad(model, torch.zeros(16, 10))
```
